File: iOS/PanoAudioChat/PanoAudioChatSDK/Source/Sections/UserList/PanoInviteListViewController.m
//
//  PanoInviteListViewController.m
//  PanoAudioChatSDK
//
//  Created by pano on 2021/7/16.
//

#import "PanoInviteListViewController.h"
#import "PanoMicInfo.h"
#import "PanoUserListCell.h"
#import <Masonry/Masonry.h>


@interface PanoInviteListViewController ()<UITableViewDataSource, UITableViewDelegate, PanoUserDelegate>

@property (nonatomic, strong) UITableView *tableView;

@end

@implementation PanoInviteListViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [PanoClientService.service.userService addDelegate:self];
}


- (void)initViews {
    _tableView = [[UITableView alloc] initWithFrame:CGRectZero style:UITableViewStylePlain];
    [_tableView registerClass:[PanoUserListCell class] forCellReuseIdentifier:@"cellID"];
    _tableView.tableFooterView = [UIView new];
    _tableView.dataSource = self;
    _tableView.delegate = self;
    _tableView.separatorStyle = UITableViewCellSeparatorStyleNone;
    _tableView.rowHeight = 55;
    [self.view addSubview: _tableView];
    self.title = NSLocalizedString(@"选择成员", nil);
}

- (void)initConstraints {
    [_tableView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.edges.mas_equalTo(self.view);
    }];
}

- (void)initService {
    _dataSource = [NSMutableArray array];
    [self reloadViews];
}

- (void)reloadViews {
    [_dataSource removeAllObjects];
    for (PanoUserInfo *user in [PanoClientService.service.userService allUserExceptMe]) {
        if (![self.onlineUser containsObject:user]) {
            [_dataSource addObject:user];
        }
    }
    [_tableView reloadData];
}

#pragma mark -- UITableViewDataSource
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return _dataSource.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    PanoUserListCell *cell = [tableView dequeueReusableCellWithIdentifier:@"cellID" forIndexPath:indexPath];
    PanoUserInfo *info = [self cellForUserInfo:indexPath];
    if (info) {
        cell->nameLabel.text = info.userName;
    }
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:true];
    PanoUserInfo *user = [self cellForUserInfo:indexPath];
    if (user) {
        _micInfo.userId = user.userId;
        [PanoClientService.service.signalService sendInviteMsg:_micInfo];
        [self dismiss];
    }
}

- (PanoUserInfo *)cellForUserInfo:(NSIndexPath *)indexPath {
    if (indexPath.row < _dataSource.count) {
        return _dataSource[indexPath.row];
    }
    return nil;
}

- (void)onUserAdded:(PanoUserInfo *)user {
    [self reloadViews];
}

- (void)onUserRemoved:(PanoUserInfo *)user {
    [self reloadViews];
}

- (void)dealloc {
    [PanoClientService.service.userService removeDelegate:self];
}

@end

File: iOS/PanoAudioChat/PanoAudioChat/Classes/Sections/UserList/PanoInviteListViewController.swift
//
//  PanoInviteListViewController.swift
//  PanoAudioChat
//
//  Created by wenhua.zhang on 2020/10/29.
//

import UIKit

class PanoInviteListViewController: PanoBaseViewController, UITableViewDataSource, UITableViewDelegate, PanoUserDelegate {
    
    var micInfo : PanoMicInfo!
    
    var onlineUser: [PanoUser]!
    
    var tableView : UITableView!
    
    var dataSource: [PanoUser]!
    
//    var footerView : PanoUserListFooterView!
    var emptyView: UILabel!
    
    deinit {
        PanoUserService.service()?.delegates.removeDelegate(self)
    }
    override func viewDidLoad() {
        super.viewDidLoad()
        PanoUserService.service()?.delegates.addDelegate(self)
    }
    
    override func initViews() {
        tableView = UITableView()
        tableView.register(PanoUserListCell.self, forCellReuseIdentifier: "cellID")
        tableView.tableFooterView = UIView()
        tableView.dataSource = self
        tableView.delegate = self
        tableView.separatorStyle = .none
        tableView.rowHeight = 55
        self.view.addSubview(tableView)
        
        self.title = NSLocalizedString("选择成员", comment: "")
    }
    
    override func initConstraints() {
        tableView.snp.makeConstraints { (make) in
            make.left.top.right.bottom.equalToSuperview()
        }
    }
    
    override func initServices() {
        reloadViews()
    }
    
    func reloadViews() {
        dataSource = PanoUserService.service()?.allUserExceptMe();
        print("\(String(describing: dataSource))")
        dataSource.removeAll { (user ) -> Bool in
            self.onlineUser.contains(user)
        }
        print("\(String(describing: dataSource))")
        tableView.reloadData()
    }
    
    // MARK: UITableViewDataSource
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        dataSource.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "cellID", for: indexPath) as! PanoUserListCell
        if let info = cellForUserInfo(indexPath: indexPath) {
            cell.nameLabel.text = info.userName
        }
        return cell
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        if let user = cellForUserInfo(indexPath: indexPath) {
            micInfo.userId = user.userId
            _ = PanoClientService.service().signalService?.sendInviteMsg(micInfo: micInfo)
            self.dismiss()
        }
    }
    
    func cellForUserInfo(indexPath: IndexPath) -> PanoUser? {
        let micInfos = dataSource
        if indexPath.row < micInfos?.count ?? 0 {
            let info = micInfos?[indexPath.row]
            return info
        }
        return nil
    }
    
    // MARK: PanoUserDelegate
    func onUserAdded(user: PanoUser) {
        reloadViews()
    }
    
    func onUserRemoved(user: PanoUser) {
        reloadViews()
    }
    
    func onUserPropertyChanged(user: PanoUser?) {
        
    }
}
